INA237 Sensor Bug Fixes and Enhancements (with unit tests)

EricNRS opened the pull request:
I have fixed several bugs in the INA237 and INA230 sensor drivers and also added unit tests to make sure the driver doesn't get broken in the future and/or make it easier to fix.

Fixes #60722

To run unit tests:
```sh
./scripts/twister -vv -p native_posix -T tests/drivers/sensor/ina237
./scripts/twister -vv -p native_posix -T tests/drivers/sensor/ina230
```

EricNRS committed:
drivers: sensor: ina237: add shunt voltage support

Add ability to retrieve the shunt voltage if the end user wants to do
the power calculation manually to handle negative power values.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
test: sensor: ina237: add emulator unit test

Add emulator unit test of the INA237.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
dt-bindings: sensor: ina237: add configuration properties

Add properties to replace the configuration register values.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
drivers: sensor: ina237: add high-precision mode

The current-shunt calibration requires a factor of 4x if high-precision
mode is selected.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -63,7 +63,7 @@ static int ina237_channel_get(const struct device *dev, enum sensor_channel chan
 	case SENSOR_CHAN_POWER:
 		/* see datasheet "Current and Power calculations" section */
-		power_uw = (data->power * INA237_POWER_SCALING * config->current_lsb) / 10000U;
+		power_uw = (data->power * INA237_POWER_SCALING * config->current_lsb) / 10U;
```
Since data->power is 24 bits and config->current_lsb can be up to 16-bits, this will still overflow for larger current LSBs.  Will add a test and update this today.

EricNRS replied:
Fixed and added two new DT test configurations to catch this overflow.

bbilas commented:
@bperseghetti @xyzzy42 @gmarull PTAL

gmarull commented on drivers/sensor/ina23x/ina237_emul.c:
```diff
@@ -0,0 +1,173 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Emulator for the TI INA237 I2C power monitor
```
sad to see emulators being used to mock. why can't we just use FFF?

EricNRS replied:
I had followed the pattern of some of the other sensors which did emulators.  Also, in this case part of the complexity is the calculation done at compile time using the DT configuration for the struct `ina237_config::cal` property  in [drivers/sensor/ina23x/ina237.c - INA237_DRIVER_INIT](https://github.com/zephyrproject-rtos/zephyr/pull/60717/commits/3e24cddf54de848d38554f692598a048fa67ff6a#diff-9b554dcbdeedb8966d7d2168b09914d60c3dbc77afbf0d0c936d73f5c3b5f4f5).

The FFF could potentially be used to replace the `ina23x_reg_read*()' and 'ina23x_reg_write()` functions if the rest of the functionality works.  Do you have a suggestion on best to replace those functions that are in `zephyr/drivers/sensor/ina23x/ina23x_common.c`?  I am used to other mocking frameworks and have limited experience with FFF.  If we can replace the `ina23x_common.c` functions, then that would allow eliminating the I2C message decode/encode portion of the `ina237_emul.c` file.  The register array would probably still be used since that is the easiest way to capture multiple writes done by `ina237_init()`.

gmarull replied:
I think it's better to fix the issue for now and open another PR to discuss testing approach.

EricNRS replied:
I tried to use FFF to replace the `ina23x_reg_read*()` and `ina23x_reg_write()` functions and I wasn't able to solve the problem where the INA237 driver did reads and writes before the unit test started and I could set the .custom_fake to the mocked function.  Not sure if there is a way to load the driver dynamically and if that is a possibility for testing.

I did do another version where I did link-time replacement of the `ina23x_reg_read*()` and `ina23x_reg_write()` functions with mocks.  It avoids have to emulate individual I2C transactions and doesn't create a specific emulator, but there is still a generic I2C emulator created behind the scenes.  It isn't any better than the emulated version and has the disadvantage of testing less.  You can take a look here:  [Version using ina23x_reg_readXX() and ina23x_reg_write() mocks](https://github.com/zephyrproject-rtos/zephyr/commit/a0617c227bda35141abdbd96b023eb910a56a2ee)

Taking a look at other options, the `zephyr/tests/drivers/sensor/icm42688` test case is interesting.  It uses FFF, but just for the interrupt mock and still uses an emulator.

Here are all of the sensor tests that I am aware of and a summary of what they are doing:

```
tests/drivers/sensor
├── accel                      I2C and SPI emulators behind the scenes, but doesn't test much
├── akm09918c            Uses I2C emulator in ./drivers/sensor/akm09918c/akm09918c_emul.c
├── generic                  Uses DEVICE_DEFINE() to create a dummy sensor_driver_api driver
├── icm42688              Uses SPI emulator in ./drivers/sensor/icm42688/icm42688_emul.c
├── ina237                   The driver we are talking about
└── sbs_gauge             I2C driver (but don't see any reference to an emulator).  Focuses on making sure driver simply
                                       returns a value vs. actually doing full testing.
```

So at this point, what would you like to do.  Should we:
1. merge this patch set as is and open a discussion in regards to the emulator
2. just fix the bugs and remove the unit tests and open a discussion
3. some other alternative

My preference is to merge this as is given that it follows the existing patterns.  If there is an easier way to test it, I am all for that as well since it will be useful in the future, but after spending an entire day trying different options, I'm still convinced the first approach here is the best as the alternatives test less and are just as complicated.

EricNRS replied:
@gmarull - please take another look and let me know if you are happy with this as-is or would like to refactor it.

EricNRS replied:
I moved the emulator out of the drivers/sensor/ina23x and included it with the test case.  Hopefully the cluttering of the drivers folder with the emulators was your concern and you don't mind then in the emulator.

Changes will be included in the next push to the PR.

xyzzy42 commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -25,13 +25,20 @@ LOG_MODULE_REGISTER(INA237, CONFIG_SENSOR_LOG_LEVEL);
+ * @brief Scale die temperture from 0.125 degC/bit to micro-degrees C
+ *  Note that the bottom 4 bits are reserved and are always zero.
+ */
+#define INA237_DIETEMP_TO_uDegC(x) ((x >> 4) * 125000)
```
There should be `()` around the `x` in the macro.  Otherwise consider what happens if someone writes  `INA237_DIETEMP_TO_uDegC(data->die_temp & ~0x000F)` or something like that.

EricNRS replied:
Good catch, thank you.

EricNRS replied:
Fixed, will be in next patchset.

xyzzy42 commented on include/zephyr/dt-bindings/sensor/ina237.h:
The more "device tree" way to do this would be to have individual properties in the bindings, like `high-precision;`,  or `vbus-conv-time-us = <1052>;`  It's not difficult to convert those into the proper config register values for the device config struct at compile time.

E.g.:
```
.adc_config = DT_INST_ENUM_IDX_OR(inst, vbus_conv_time_us, 5) << 9 |
              DT_INST_ENUM_IDX_OR(inst, vshunt_conv_time_us, 5) << 6 | ... ,
```

The dt binding system lets ENUM_IDX return the index number from a list of allowed values used in the bindings.

EricNRS replied:
That looks much better.

EricNRS replied:
Fixed, will be in next patchset.

EricNRS commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -165,6 +183,14 @@ static int ina237_read_data(const struct device *dev)
 		}
 	}
 
+	if ((data->chan == SENSOR_CHAN_ALL) || (data->chan == SENSOR_CHAN_VSHUNT)) {
```
> Is there a real use case for reading the shunt voltage?

Very true, it is just for debug at this point, although I have considered doing the math in the micro instead just as you mentioned later.  I'll remove it from the SENSOR_CHAN_ALL group.  Based upon the extra cached value, did you want me to actually remove this completely?

> TBH, until someone ports Zephyr to the 8051, all the conversion and math stuff this chip does is obsolete. The most efficient driver would just read shunt voltage and voltage, ignore the cal stuff totally, and do all the math on the CPU

Yes, I agree with you on this part after going through everything.  I also looked at doing a bulk read of all of the registers so at least there is only one I2C transaction, but that is not supported by the chip :(

EricNRS replied:
Removed it from the ALL_CHANNELS list in the next patchset.

EricNRS replied:
@xyzzy42 Please take another look and resolve this thread if you are happy with removing the shunt voltage from ALL_CHANNELS.

xyzzy42 replied:
What I think doesn't really matter, but I'd probably make it optional via a menuconfig entry.  I wouldn't want to enable it by default, since it has a continuous runtime cost with zero benefit to virtually all users.  But it could be useful for debugging or some unusual use case like non-constant shunt resistance value or using the INA237 as a mV meter.  And I do like to expose every feature I can.

Having it there, but not in the ALL_CHANNELS list seems like it would be confusing.  I don't think there's a precedent for that and it's contrary to the sensor docs.

EricNRS replied:
@xyzzy42 I have added the KConfig and kept the current shunt in the ALL_CHANNEL group.  Please take another look when you have a chance.

EricNRS commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -71,6 +71,24 @@ static int ina237_channel_get(const struct device *dev, enum sensor_channel chan
+			current_ua = (current_ua * 125) / 100;
+		} else {
+			/* standard-resolution - 5 uV/bit */
+			current_ua = (current_ua * 500) / 100;
```
Good point, will change this to just 5.

EricNRS replied:
Fixed, will be in next patch set.

EricNRS commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -63,7 +64,7 @@ static int ina237_channel_get(const struct device *dev, enum sensor_channel chan
 	case SENSOR_CHAN_POWER:
 		/* see datasheet "Current and Power calculations" section */
-		power_uw = (data->power * INA237_POWER_SCALING * config->current_lsb) / 10000U;
+		power_uw = (data->power * INA237_POWER_SCALING * config->current_lsb) / 10U;
```
> There is no reason to make power_uw 64 bit here. It won't overflow until the power reaches 4.29 kW, which is far more than this sensor is designed to measure.

There is no fixed limit that the sensor measures since the INA237 uses an external shunt resistor, so the measurement limit is just based upon the data types involved.  Power is 24 bits, INA237_POWER_SCALING is 1 bit, and current_lsb can be up to 16 bits for a total of 41 bits in the numerator.

> It can be improved by changing the … * 2 / 10 into … / 5.

Good point, will change INA237_POWER_SCALING to 5, use it as a divisor, and will adjust the name.

EricNRS replied:
Fixed, will be in next patchset.

EricNRS replied:
@xyzzy42 PTAL

EricNRS commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -49,13 +49,9 @@ static int ina237_channel_get(const struct device *dev, enum sensor_channel chan
 		break;
 
 	case SENSOR_CHAN_CURRENT:
-		if (data->current & INA23X_CURRENT_SIGN_BIT) {
```
Thanks for the Compiler Explorer link.  Probably just go with abs() in the negative case given the results.

EricNRS replied:
> Doesn't ina230 have this same problem?
Yes, it does have the same issue code issue, but I do not have INA230 hardware to test, so I didn't want to change the code and release it untested.

EricNRS committed:
drivers: sensor: ina237: fix current sign issue

Fix current sign extension logic and consolidate sensor scaling code
into a single block.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS commented:
I have updated the shunt voltage to have a reporting sensor value of mV to avoid loss of precision at low current levels since high-precision mode is 1.25 uV/bit.  This changes the val2 sensor value into nano-volts instead of micro-volts.

I have not rebased the changes so you can still do a compare.

Please have another look.  Once everything has been approved I can do a final rebase.

xyzzy42 commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -20,51 +20,69 @@ LOG_MODULE_REGISTER(INA237, CONFIG_SENSOR_LOG_LEVEL);
+{
+	val->val1 = value_microX / 1000000U;
+	val->val2 = value_microX % 1000000U;
+	if (value_microX < 0) {
```
*unsigned* 64-bit int `value_microX` is unlikely to be less than zero.

EricNRS replied:
Ooops, a bit of late night-copy-and-paste didn't go over well.  Good catch, will remove the check.

xyzzy42 commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -20,51 +20,69 @@ LOG_MODULE_REGISTER(INA237, CONFIG_SENSOR_LOG_LEVEL);
+
+static void micro_s32_to_sensor_value(struct sensor_value *val, int32_t value_microX)
+{
+	val->val1 = value_microX / 1000000L;
+	val->val2 = abs(value_microX) % 1000000L;
+	if (value_microX < 0) {
+		val->val2 = -val->val2;
+	}
```
This can be simplified and made about twice as fast if written as:
```c
    val->val1 = value_microX / 1000000L;
    val->val2 = value_microX % 1000000L;
```

The compiler can do one signed division this way, instead of an unsigned division and a signed division.

This will always generate the correct value for C99 or later.  Zephyr is C11 minimum.  In C89, it may or may not generate the correct output, depending on the CPU, but what you have now will also be wrong for the same reason.

In C99, division that produces a negative result rounds toward zero, e.g. -1200000 / 100000 = -1.  In C89, it is implementation defined if it rounds toward zero or rounds down, e.g. -1200000 / 100000 = -2.

So in C89, **val1** could have the wrong result (-2, rather than -1).  But thankfully it's been 24 years since that was addressed and we don't have to worry about it anymore.

In all C standards, `%` is defined in terms of division:  `(a / b) * b + (a % b) = a`

With `(a / b)` rounding toward zero, the result of `(a % b)` will need to be negative, e.g. -200000 in the above example, which is what we want.  If division rounded down, then it would need to be positive, 800000.

So your code does produce the correct `val2` even on the (unusual!) rounding-down C89 compiler, but it only makes a difference when `val1` will be incorrect, so the total result will still be wrong.

EricNRS replied:
Thanks the for explanation, will fix this.

EricNRS replied:
All fixed in the latest commit.  Thanks again for the detailed explanation since I have been forever thinking of % in C as modulus, so I learned something valuable from you.

MaureenHelm commented on dts/bindings/sensor/ti,ina237.yaml:
```diff
@@ -15,13 +15,80 @@ compatible: "ti,ina237"
+    type: int
+    description: |
+      Vbus conversion time in microseconds.
+    default: 1052
```
Property defaults need a justification in the description. Is this register reset value?

EricNRS replied:
Changes will be added in the next push to the PR.

EricNRS commented on dts/bindings/sensor/ti,ina237.yaml:
```diff
@@ -15,13 +15,80 @@ compatible: "ti,ina237"
 include: ti,ina23x-common.yaml
 
 properties:
+  config:
+    type: int
+    deprecated: true
+    default: 0x0000
+    description: |
+      Value of the configuration register
+      e.g shunt voltage and bus voltage ADC conversion
+      times and averaging, operating mode for INA230 or
+      delay for initial ADC conversion, shunt full scale range
+      for INA237.
+
+      Deprecated, please use new properties instead.
```
That brings up the other point that the INA237 and INA230 are really different chips that do not share common configuration values or features other than the original.  For example, the INA230 does not have a temperature sensor, the adc-mode settings are different, and the conversion times are 140, 204, 332, 588, 1100, 2116, 4156, and 8244 us.

Taking a look at the common code between the two (ina23x_common.h,  ina23x_common.c), it really is just the I2C register read/write functions that are common:
```c
int ina23x_reg_read_24(const struct i2c_dt_spec *bus, uint8_t reg, uint32_t *val);
int ina23x_reg_read_16(const struct i2c_dt_spec *bus, uint8_t reg, uint16_t *val);
int ina23x_reg_write(const struct i2c_dt_spec *bus, uint8_t reg, uint16_t val);
```

It seems like the best option here is to get rid of the ina23x* code and create completely independent ina230 and ina237 drivers.  Maybe that is something we could open up as a "Good first issue" task instead of throwing it into this PR?

EricNRS replied:
I added the DT configurations to the INA230 to address your concern.  I also added a unit test to make sure I didn't break anything.  Changes will be in the next push to the PR.

EricNRS replied:
@MaureenHelm - please take another look when you have a chance.

EricNRS committed:
dt-bindings: sensor: ina230: add configuration properties

Add properties to replace the configuration register value.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
test: sensor: ina230: add emulator unit test

Add emulator unit test of the INA230.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS committed:
drivers: sensor: ina230: fix current sign issue

Fix sign handling for ina230 current calculation.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS commented:
I have hopefully addressed all of the review comments and now added an INA230 unit test and fixes for issues found the INA237 driver.  Thanks for all of comments and your time spent reviewing the changes.

MaureenHelm commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -26,6 +26,14 @@ config INA237
+	default y
+	depends on DT_HAS_TI_INA237_ENABLED
+	help
+	  Enable VShunt measurement for INA237.  This requires an extra I2C
+	  read when enabled and
```
and what?

EricNRS replied:
Ooops, thanks for catching that.  Will get it updated today.

EricNRS replied:
Updated the description and pushed the commit with just the change.  Will push a rebase next.

bperseghetti reviewed (dismissed):
Thanks for your work on this. We use the INA226 but due to it's [similarity to the INA230](https://www.ti.com/compare-products/?id=3174&type=GPT&mode=alternate-gpn&partList=INA226,INA230&sticky=true) we have so far just used the INA230 for our code base.

EricNRS commented:
Finished with changes based upon review comments.  Next push will be a rebase.

bbilas commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -26,6 +26,15 @@ config INA237
 
+config INA237_VSHUNT
+	bool "INA237"
+	default n
```
It's superfluous, please remove it.

bbilas commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -26,6 +26,15 @@ config INA237
 	  Enable driver for INA237.
 
+config INA237_VSHUNT
+	bool "INA237"
```
It seems that it is copy&paste from the above. Please use a more suitable string.

bbilas commented on dts/bindings/sensor/ti,ina230.yaml:
```diff
@@ -16,6 +16,58 @@ compatible: "ti,ina230"
+      Value of the configuration register
+      e.g shunt voltage and bus voltage ADC conversion
+      times, ADC and averaging, and ADC operating mode.
+  alert-config:
```
Please add `nl` before `alert-config`.

bbilas commented on include/zephyr/dt-bindings/sensor/ina237.h:
```diff
@@ -9,6 +9,7 @@
 #include <zephyr/dt-bindings/dt-util.h>
 
 /* Operating Mode */
+#define INA237_CFG_HIGH_PRECISION	BIT(4)
```
Please align it with the rest below.

EricNRS replied:
Fixed in latest push.

bbilas commented on tests/drivers/sensor/ina230/src/ina230_emul.c:
```diff
@@ -0,0 +1,172 @@
+	.transfer = ina230_emul_transfer_i2c,
+};
+
+#define INA230_EMUL(n)                                                                 \
```
Would you mind aligning the `\` with the rest, please?

EricNRS replied:
Retabbed in the last version pushed.

bbilas commented on tests/drivers/sensor/ina230/src/ina230_emul.h:
```diff
@@ -0,0 +1,13 @@
+ *
+ * Emulator for the TI INA230 I2C power monitor
+ */
+#pragma once
```
More common is:

<code>
#ifndef HEADERNAME_H
#define HEADERNAME_H
</code>

EricNRS replied:
I thought I saw it more in the codebase, but checking it looks like there are only 8 other instances.  Have replaced with standard header guards in the latest push.

bbilas commented on tests/drivers/sensor/ina237/src/ina237_emul.c:
```diff
@@ -0,0 +1,174 @@
+};
+
+#define INA237_EMUL(n)                                                                             \
+	static const struct ina237_emul_cfg ina237_emul_cfg_##n = {                        \
```
ditto

EricNRS replied:
Fixed in latest push.

bbilas commented on tests/drivers/sensor/ina237/src/ina237_emul.h:
```diff
@@ -0,0 +1,14 @@
+ * Emulator for the TI INA237 I2C power monitor
+ */
+
+#pragma once
```
ditto

EricNRS replied:
Fixed in latest push.

xyzzy42 commented on drivers/sensor/ina23x/Kconfig:
```diff
@@ -26,6 +26,15 @@ config INA237
+	help
+	  Enable direct shunt voltage measurement for INA237. This requires an extra I2C
+	  read when enabled and SENSOR_CHAN_ALL is selected, so only enable if the shunt
+	  voltage measurement is required.
```
I might add something like:
> The Vshunt measurement is the same ADC reading as the current measurement.  The only difference is that the conversion from voltage to current with the configured `cal` value is not done.  Thus it is redundant with current and not enabled by default.

MaureenHelm commented:
Please have a look at the failing twister test:
```
INFO    - The following issues were found (showing the top 10 items):
INFO    - 1) tests/drivers/sensor/ina237/drivers.sensor.ina237 on native_posix error (Build failure)
```

EricNRS commented:
> Please have a look at the failing twister test:
> 
> ```
> INFO    - The following issues were found (showing the top 10 items):
> INFO    - 1) tests/drivers/sensor/ina237/drivers.sensor.ina237 on native_posix error (Build failure)
> ```

Thanks @MaureenHelm and sorry about that.  Didn't name the header guard well.  All fixed and runs locally, so should be good on the CI server now with the last push.  Will do another quick rebase and push that.

xyzzy42 commented on drivers/sensor/ina23x/ina237.h:
```diff
@@ -43,6 +43,7 @@ struct ina237_data {
 	uint16_t bus_voltage;
 	uint32_t power;
 	int16_t die_temp;
+	int16_t shunt_voltage;
```
This can be ifdef protected, so it takes take space in the struct unless enabled.

EricNRS replied:
Good point, done and pushed.
